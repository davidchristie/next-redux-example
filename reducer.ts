import { actionTypes } from "./actions";

export interface RootState {
  count: number;
  error: Error | undefined;
  lastUpdate: number;
  light: boolean;
  placeholderData: any | null;
}

export const exampleInitialState: RootState = {
  count: 0,
  error: undefined,
  lastUpdate: 0,
  light: false,
  placeholderData: null
};

function reducer(state: RootState = exampleInitialState, action) {
  switch (action.type) {
    case actionTypes.FAILURE:
      return {
        ...state,
        ...{ error: action.error }
      };

    case actionTypes.INCREMENT:
      return {
        ...state,
        ...{ count: state.count + 1 }
      };

    case actionTypes.DECREMENT:
      return {
        ...state,
        ...{ count: state.count - 1 }
      };

    case actionTypes.RESET:
      return {
        ...state,
        ...{ count: exampleInitialState.count }
      };

    case actionTypes.LOAD_DATA_SUCCESS:
      return {
        ...state,
        ...{ placeholderData: action.data }
      };

    case actionTypes.TICK_CLOCK:
      return {
        ...state,
        ...{ lastUpdate: action.ts, light: !!action.light }
      };

    default:
      return state;
  }
}

export default reducer;
