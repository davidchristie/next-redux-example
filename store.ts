import { Action, applyMiddleware, createStore, Middleware } from "redux";
import createSagaMiddleware, { Task } from "redux-saga";
import rootReducer, { exampleInitialState, RootState } from "./reducer";
import { rootSaga } from "./sagas";

const bindMiddleware = (middleware: Middleware[]) => {
  if (process.env.NODE_ENV !== "production") {
    const { composeWithDevTools } = require("redux-devtools-extension");
    return composeWithDevTools(applyMiddleware(...middleware));
  }
  return applyMiddleware(...middleware);
};

export const createReduxStore = (initialState: Partial<RootState> = {}) => {
  const preloadedState = {
    ...exampleInitialState,
    ...initialState
  };
  const sagaMiddleware = createSagaMiddleware();
  const store = createStore<
    RootState,
    Action<unknown>,
    { sagaTask: Task },
    unknown
  >(rootReducer, preloadedState, bindMiddleware([sagaMiddleware]));
  store.sagaTask = sagaMiddleware.run(rootSaga);
  return store;
};
