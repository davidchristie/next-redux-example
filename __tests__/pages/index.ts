import { NextPageContext } from "next";
import IndexPage from "../../pages/index";
import {
  describeComponent,
  itShouldDispatchAction,
  itShouldMatchSnapshot,
  itShowsText,
  mockDispatch
} from "../../test-utils";
import { startClock } from "../../actions";
import { exampleInitialState } from "../../reducer";
import { createReduxStore } from "../../store";

describeComponent(IndexPage, {}, getContext => {
  itShowsText(getContext, "Index Page");

  itShouldMatchSnapshot(getContext);

  itShouldDispatchAction(getContext, startClock());

  describe("Initial Props", () => {
    it("Should match snapshot", async () => {
      const initialProps = await IndexPage.getInitialProps!({
        AppTree: () => null,
        ctx: {
          store: createReduxStore(),
          isServer: false
        },
        pathname: "/",
        query: {}
      } as NextPageContext);
      expect(initialProps).toMatchSnapshot();
    });

    describe("If data has already been loaded", () => {
      it("Should match snapshot", async () => {
        const initialState = {
          ...exampleInitialState,
          placeholderData: [{ id: 1 }, { id: 2 }, { id: 3 }]
        };
        const store = createReduxStore(initialState);
        mockDispatch(store);
        const initialProps = await IndexPage.getInitialProps!({
          AppTree: () => null,
          ctx: {
            store,
            isServer: false
          },
          pathname: "/",
          query: {}
        } as NextPageContext);
        expect(initialProps).toMatchSnapshot();
      });
    });
  });
});
