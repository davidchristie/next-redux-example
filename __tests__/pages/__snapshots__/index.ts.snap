// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IndexPage Component Initial Props If data has already been loaded Should match snapshot 1`] = `
Object {
  "isServer": false,
}
`;

exports[`IndexPage Component Initial Props Should match snapshot 1`] = `
Object {
  "isServer": false,
}
`;

exports[`IndexPage Component Should match snapshot 1`] = `
<DocumentFragment>
  <div>
    <header>
      <a
        href="https://github.com/davidchristie/next-redux-example"
      >
        GitHub
      </a>
    </header>
    <h1>
      Index Page
    </h1>
    <div
      class="jsx-1453841751 "
    >
      00:00:00
    </div>
    <div
      class="jsx-1118913787"
    >
      <h1
        class="jsx-1118913787"
      >
        Count: 
        <span
          class="jsx-1118913787"
        >
          0
        </span>
      </h1>
      <button
        class="jsx-1118913787"
      >
        +1
      </button>
      <button
        class="jsx-1118913787"
      >
        -1
      </button>
      <button
        class="jsx-1118913787"
      >
        Reset
      </button>
    </div>
    <nav>
      <a
        href="/other"
      >
        Navigate: Other Page
      </a>
    </nav>
  </div>
</DocumentFragment>
`;
